### ISSUE TITLE
Storybook - Storybook is displaying a white page upon opening + console errors

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#40041]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___
issue found when validating #39372
**Version Number:** 1.4.62-0
**Reproducible in staging?:** y
**Reproducible in production?:** no
**If this was caught during regression testing, add the test name, ID and link from TestRail:** n/a
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** applause internal team
**Slack conversation:**

## Action Performed:
1. Navigate to https://staging.new.expensify.com/docs/index.html

## Expected Result:
Page loads

## Actual Result:
Storybook is displaying a white page upon opening. "Refused to execute inline script" console errors are showing.
## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [x] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/38435837/8c76be2d-8a60-48d7-b86e-a

[Bug6444964_1712776318500!staging.new.expensify.com-1712775989936.txt](https://github.com/Expensify/App/files/14937606/Bug6444964_1712776318500.staging.new.expensify.com-1712775989936.txt)
0cfba8e5b3c

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by pecanoro** [comment]
Managing this here: https://github.com/Expensify/App/pull/39372#issuecomment-2049436778

**Comment by alexpensify** [comment]
I think we should close this one in favor of https://github.com/Expensify/App/pull/39372#issuecomment-2049436778 but @marcaaron let me know if that's the wrong action. Thanks!

**Comment by marcaaron** [comment]
Works for me. Thanks!

**Comment by pecanoro** [comment]
Going to reopen it since it's still broken. I am assigning @allroundexperts since he created the original PR so we can look into how this is happening

**Comment by alexpensify** [comment]
@allroundexperts any update here?

**Comment by allroundexperts** [comment]
This seems to be related to the backend.

More context: https://expensify.slack.com/archives/C049HHMV9SM/p1714094217386239?thread_ts=1714064859.296509&cid=C049HHMV9SM

**Comment by marcaaron** [comment]
I'm not sure the nonce solution is going to work for us as we use one service to generate the nonce and another to serve the HTML. Maybe we can explore bundling the JS instead of having it load in script tags? I think this could be achieved externally so gonna open this up for more eyes while we discuss any alternatives internally.

**Comment by sunstar21** [comment]
Hello - I submitted a proposal on Upwork to fix this bug. I have dealt with this issue before and would love to help fix it.

**Comment by sunstar21** [comment]
Contributor details
Your Expensify account email: usertesting190@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01c79972a54a7a5990

**Comment by oscarguinane** [comment]
Hey guys, fixed:

<img width="1524" alt="Screen Shot 2024-05-01 at 8 44 12 PM" src="https://github.com/Expensify/App/assets/4477702/3414d716-a2ea-4687-ab4a-8df09686258a">

**Comment by oscarguinane** [comment]
Contributor details
Your Expensify account email: oscar+expensify@bloodandtreasure.com
Upwork Profile Link: https://www.upwork.com/freelancers/~017275408a967b697f

**Comment by oscarguinane** [comment]
Fixed

Contributor details
Your Expensify account email: [oscar+expensify@bloodandtreasure.com](mailto:oscar+expensify@bloodandtreasure.com)
Upwork Profile Link: https://www.upwork.com/freelancers/~017275408a967b697f

**Comment by oscarguinane** [comment]
How about now?

Contributor details
Your Expensify account email: oscar+expensify@bloodandtreasure.com
Upwork Profile Link: https://www.upwork.com/freelancers/~017275408a967b697f

**Comment by oscarguinane** [comment]
Literally all I needed to do to fix the issue was add this to the CSP header: `script-src 'self' 'unsafe-eval' https://staging.new.expensify.com;`

**Comment by oscarguinane** [comment]
Ok well... let me know next steps, happy to create a PR...

**Comment by sunstar21** [comment]
Yeah I said this in my proposal..... 

"To fix this problem, you need to go to your webserver and add 'unsafe-inline' to the CSP for script-source." @oscarguinane
 
Please accept my proposal because this diagnosis took around 1 hour and I also cloned the code and fixed it on my side ready for the pull request and merging

**Comment by oscarguinane** [comment]
@sunstar21 You have not said the same thing in your proposal based on that quote, specifically you have to add the URL for staging to the web server. The staging URL doesn't show up at all and that's the issue, `unsafe-inline` is already present for a bunch of other sources.

**Comment by oscarguinane** [comment]
I have tested it locally and it's loading after I build it with the staging config

**Comment by oscarguinane** [comment]
Let me know what else you need from me

**Comment by allroundexperts** [comment]
Hi @sunstar21 & @oscarguinane.

Please follow the [proposal template](https://github.com/Expensify/App/blob/main/contributingGuides/PROPOSAL_TEMPLATE.md) for posting proposals.

**Comment by allroundexperts** [comment]
> I have tested it locally and it's loading after I build it with the staging config

@oscarguinane Can you confirm that it does not work locally after you build with staging config without your changes? Can you mention the steps you are using to test?

**Comment by sunstar21** [comment]
@allroundexperts I sent a proposal last week (after reading the proposal guidelines) on upwork containing a diagnosis of the issue and how to fix it

**Comment by allroundexperts** [comment]
@sunstar21 We require a proposal here. Not on upwork.

**Comment by oscarguinane** [comment]
Ok that makes sense, here's mine:

### Please re-state the problem that we are trying to solve in this issue.

The page is blank on staging when viewing the storybook docs. When you load the page or reload the page it stays blank.

### What is the root cause of that problem?

The root cause of the problem is that there is no reference to the staging URL for the CSP (content security policy) header. This means that although a bunch of other endpoints are covered, running scripts directly on staging does not work so the page doesn't execute any import, not javascript is executed, and the single application page does not render.

### What changes do you think we should make in order to solve the problem?

We need to include the staging header in the CSP to allow scripts to to be executed.

This is what the CSP was on staging:

```
<meta http-equiv="Content-Security-Policy" content="script-src https://kit.fontawesome.com https://cdnjs.cloudflare.com https://www.google.com https://www.gstatic.com/ https://cse.google.com http://cse.google.com https://partner.googleadservices.com/ https://cdnjs.cloudflare.com/ajax/libs/tocbot/ https://www.googletagmanager.com 'self' 'unsafe-eval' 'unsafe-inline'; style-src cdnjs.cloudflare.com cse.google.com www.google.com ka-f.fontawesome.com fonts.googleapis.com 'self' 'unsafe-inline'" >
```

It's missing any reference to staging, this is what it should look like on staging:

```
<meta http-equiv="Content-Security-Policy" content="script-src https://staging.new.expensify.com https://kit.fontawesome.com https://cdnjs.cloudflare.com https://www.google.com https://www.gstatic.com/ https://cse.google.com http://cse.google.com https://partner.googleadservices.com/ https://cdnjs.cloudflare.com/ajax/libs/tocbot/ https://www.googletagmanager.com 'self' 'unsafe-eval' 'unsafe-inline'; style-src cdnjs.cloudflare.com cse.google.com www.google.com ka-f.fontawesome.com fonts.googleapis.com 'self' 'unsafe-inline'" >
```

**Comment by eh2077** [comment]
@allroundexperts I also added a [comment](https://expensify.slack.com/archives/C049HHMV9SM/p1714094217386239?thread_ts=1714064859.296509&cid=C049HHMV9SM) in the internal bug report thread - I think we should fix this CSP issue using the `nonce` solution, like https://cssinjs.org/csp/?v=v10.10.0

I can write a proposal if the idea sounds good to you.

**Comment by oscarguinane** [comment]
> I have tested it locally and it's loading after I build it with the staging config
> 
> @oscarguinane Can you confirm that it does not work locally after you build with staging config without your changes? Can you mention the steps you are using to test?

@allroundexperts certainly, see this attached video. You can see that when testing with the CSP, if the header is not specified for the local domain, the page does not load, and when we add the domain, it does load.

I tried with the PR to set the meta tag for the CSP but it looks like the default layout is not used as a container for storybook so it just needs to be updated on AWS.

https://github.com/Expensify/App/assets/4477702/0adbf435-8d54-4a22-b8b3-73e6c04dda5c

**Comment by sunstar21** [comment]
@allroundexperts 

Alright, here is my proposal:

**Restating the problem**:
When you navigate to this link https://staging.new.expensify.com/docs/index.html the page doesn't render and an error is thrown in the console

**What is the root cause of the problem?** 
This error occurs when you include a script inside your webpage, that isn't referenced from the website itself. To fix this problem, you need to go to your webserver and add 'unsafe-inline' to the CSP for script-source. You can also add a reference to the staging URL for the CSP 'https://staging.new......'  If you are unable to do this, you will need to take out all the embedded script tags and move them into an external source file. 

**What changes need to be made in order to solve this problem?**
You can do 1 of three things: Add this line 'unsafe-inline' to the CSP for the script-source, you can add a reference to the staging URL for the CSP 'https://staging.new.......', or you can remove all the embedded script tags and move them to an external source file

**Comment by sunstar21** [comment]
@oscarguinane @allroundexperts 

Oscar, you have a great solution, but you don't want to deploy staging hardcoded lines of code to production. I have a better more permanent solution in mind that enviornment agnostic.

**Comment by oscarguinane** [comment]
@sunstar21 True, there's nothing in code that we can upload it seems because the configuration is totally controlled on the backend outside of the codebase anyway.

I think this issue is a little too easy too fix so if they do end up stealing my idea / solution and implement it on the AWS side, then so be it 😅

I just loaded the webpage but it's actually still broken for me so they haven't stolen anything yet. Maybe you have a CSP tester on like I showed in the video? If you turn that off, it will fail to load

**Comment by sunstar21** [comment]
@oscarguinane Ah okay - it is now broken once again. Weird...

**Comment by oscarguinane** [comment]
Interesting

**Comment by oscarguinane** [comment]
Interesting

**Comment by allroundexperts** [comment]
@marcaaron The current proposals are suggesting to just replace `script-src 'self' 'unsafe-eval'` with `script-src 'self' 'unsafe-eval' https://*.new.expensify.com` in the `CSP` header. This does not involve the nonce. Can you give it a shot?

**Comment by alexpensify** [comment]
@marcaaron any update to @allroundexperts's question? Thanks!

**Comment by marcaaron** [comment]
Unfortunately, I've not had a chance to catch up here. @roryabraham has a PR that should fix this, but we're holding off until next week due to the current freeze.

**Comment by alexpensify** [comment]
Waiting for the freeze to end

**Comment by marcaaron** [comment]
Merged an internal fix [here](https://github.com/Expensify/Cloudflare-Workers/pull/141). I believe we still need to deploy it.

**Comment by roryabraham** [comment]
that fix caused a regression. Unfortunately, there's no way to test cloudflare workers locally 😓

**Comment by marcaaron** [comment]
No update yet. I think we just need to try deploying the change again soon.

**Comment by alexpensify** [comment]
## Heads up, I will be offline until Tuesday, May 28, 2024, and will not actively watch over this GitHub during that period.

If anything urgent is needed here, please ask for help in the #expensify-open-source Slack Room-- thanks!

**Comment by alexpensify** [comment]
Catching up from being OOO, @marcaaron, we have been in a freeze. Is there a plan to try again to deploy the change this week? Thanks!

**Comment by marcaaron** [comment]
I'll check with Rory. This is pretty low priority compared to what I am currently tasked with.

**Comment by roryabraham** [comment]
agree this is pretty low priority. I don't think I'd want to try this again while "flying blind". Would love to get local testing for the NewDot CSP Cloudflare Worker set up first. Maybe we put this on HOLD for https://github.com/Expensify/App/issues/38072

**Comment by marcaaron** [comment]
Works for me!

**Comment by alexpensify** [comment]
Thanks for the feedback, I've updated this one to be on hold.

**Comment by alexpensify** [comment]
Weekly Update: Waiting on https://github.com/Expensify/App/issues/38072

**Comment by alexpensify** [comment]
Weekly Update: Same state for https://github.com/Expensify/App/issues/38072

**Comment by alexpensify** [comment]
Weekly Update: No update in https://github.com/Expensify/App/issues/38072

**Comment by alexpensify** [comment]
Weekly Update: No movement in https://github.com/Expensify/App/issues/38072

**Comment by alexpensify** [comment]
Weekly Update: No update yet in https://github.com/Expensify/App/issues/38072

**Comment by alexpensify** [comment]
Weekly Update: https://github.com/Expensify/App/issues/38072 waiting for that GH

**Comment by alexpensify** [comment]
Weekly Update: https://github.com/Expensify/App/issues/38072

**Comment by alexpensify** [comment]
Weekly Update: On hold for https://github.com/Expensify/App/issues/38072

**Comment by alexpensify** [comment]
Weekly Update: Hold https://github.com/Expensify/App/issues/38072

**Comment by alexpensify** [comment]
Weekly Update: https://github.com/Expensify/App/issues/38072

**Comment by alexpensify** [comment]
Weekly Update: Waiting https://github.com/Expensify/App/issues/38072

**Comment by alexpensify** [comment]
Weekly Update: Same

**Comment by alexpensify** [comment]
Weekly Update: On Hold

**Comment by alexpensify** [comment]
Weekly Update: https://github.com/Expensify/App/issues/38072 Holding here

**Comment by alexpensify** [comment]
Weekly Update: On Hold

**Comment by alexpensify** [comment]
Weekly Update: Hold

**Comment by alexpensify** [comment]
Weekly Update: On hold for https://github.com/Expensify/App/issues/38072

**Comment by alexpensify** [comment]
Weekly Update: https://github.com/Expensify/App/issues/38072 was moved to Monthly, so I moved this one too.

**Comment by alexpensify** [comment]
The other issue is still monthly

**Comment by alexpensify** [comment]
https://github.com/Expensify/App/issues/38072 is monthly

**Comment by alexpensify** [comment]
## Heads up, I will be offline until Wednesday, December 18, 2024, and will not actively watch over this GitHub during that period.

If this GitHub requires an urgent update, please ask for help in the #expensify-open-source Slack Room. If the inquiry can wait, I'll review it when I return online.

**Comment by alexpensify** [comment]
No movement in https://github.com/Expensify/App/issues/38072, still on hold here

**Comment by alexpensify** [comment]
I believe we are still on hold for https://github.com/Expensify/App/issues/38072

--- 

## Heads up, I will be offline until Friday, January 3, 2025, and will not actively watch over this GitHub during that period.

If this GitHub requires an urgent update, please ask for help in the #expensify-open-source Slack Room. If the inquiry can wait, I'll review it when I return online.

**Comment by alexpensify** [comment]
Weekly Update: On hold for https://github.com/Expensify/App/issues/38072

**Comment by alexpensify** [comment]
Weekly Update: Moving to monthly to match https://github.com/Expensify/App/issues/38072

**Comment by alexpensify** [comment]
Monthly Update: Waiting on https://github.com/Expensify/App/issues/38072

**Comment by alexpensify** [comment]
Monthly Update: Hold for https://github.com/Expensify/App/issues/38072

**Comment by mallenexpensify** [comment]
Removed `Help Wanted` and price in title, it doesn't look like it's External

**Comment by roryabraham** [comment]
no update, still on HOLD

**Comment by alexpensify** [comment]
On hold

**Comment by roryabraham** [comment]
still on hold

**Comment by alexpensify** [comment]
I think we are getting close since the other issue has a PR under review.

**Comment by alexpensify** [comment]
Still on hold here

**Comment by alexpensify** [comment]
On Hold

**Comment by alexpensify** [comment]
Hold

**Comment by alexpensify** [comment]
Thank you for the update!

**Comment by roryabraham** [comment]
I'm not likely to prioritize this one on this short week
